### DISCUSSION TITLE
What do the different Block Sizes do?

### DISCUSSION THREAD

**Discussion by MM502** [apohl79/audiogridder#1096] [Q&A]
I have AG running nicely, hosting Acustica plug-ins, on a dedicated server. Once everything was working well I wanted to dig into the various options to see what they did and if I might be able to squeeze more performance out of the server.

Changing the Buffer Size to 3x Block Size resulted in a big performance gain. 
![AG Buffer Size 2022-11-11 174700](https://user-images.githubusercontent.com/48056686/201399834-ec1dfc60-6ede-4fee-8144-abce7763dd81.jpg)

Going up to 4x made things much worse, however.

I then went through the Blocks option. Anything above 1 block did not work (Perf Net hit the red).
![AG 2 blocks 2022-11-11 175031](https://user-images.githubusercontent.com/48056686/201400114-8fdd2c02-0f5e-426a-ba89-75e79faf46b8.jpg)
Trying all sorts of permutations of Block Size and [number of] Blocks always resulted in inferior performance as did using fixed size output buffers and allowing individual buffer sizes by plug-in.

Nothing wrong with any of that but I feel like I am shooting in the dark. Would be grateful for explanation of the above functionality (or pointing towards it - search yielded no results).

Thank you!

**Comment by apohl79** [comment]
Most of the options are explained here: https://audiogridder.com/plugin/#settings - But some things are still missing.

AG uses a queue of audio buffers of a particular block size to buffer outbound/inbound audio. This block size is usually defined by the DAW/host (and can be configured in many hosts). And most hosts are passing audio into a plugin with this exact block size.

The block size setting (which is pretty new) lets you increase the block size AG uses. This effectively increases the audio chunks, that AG passes to the server for processing. Sometimes it can be beneficial to have larger chunks to improve latency when it is exceeding time limits. For example a plugin might need ~25ms to process 1024 samples, which is too slow for 48kHz, but it might achieve processing 2048 samples in ~35ms, which is ok.

This would also means, that the DAW has to call the AG plugin twice or more before it sends a chunk of audio to the server. This might lead to visual effects to meters or analyzers etc (they might stutter), as audio comes in larger chunks but less often. But it has no impact to the audio data.

The number of blocks define the size of the buffer queue that AG uses to send/receive audio chunks asynchronously in the background of the hosts audio thread.

**Comment by MM502** [comment]
Thank you!
